compute_tools/src/http/routes/prewarm_lfc.rs

```diff
@@ -0,0 +1,188 @@
+    }
+    spawn(async move {
+        let Err(err) = prewarm_lfc_impl(parts, &state.tokio_conn_conf).await else {
+            state.state.lock().unwrap().prewarm_state.status = Completed;
+            return;
+        };
+        error!(%err);
+        let state = &mut state.state.lock().unwrap().prewarm_state;
+        state.status = Failed;
+        state.error = err.to_string();
+    });
+    StatusCode::OK.into_response()
+}
+
+async fn prewarm_lfc_impl(parts: Parts, conn: &tokio_postgres::Config) -> anyhow::Result<()> {
+    let Parts { uri, token } = parts;
+    info!(%uri, "requesting LFC state from endpoint storage");
+
+    let request = Client::new().get(uri.clone()).bearer_auth(token);
+    let res = request.send().await.context("querying endpoint storage")?;
+    let status = res.status();
+    if status != StatusCode::OK {
+        anyhow::bail!("{status} querying endpoint storage")
+    }
+
+    let mut uncompressed = Vec::new();
+    let lfc_state = res
+        .bytes()
+        .await
+        .context("getting request body from endpoint storage")?;
+    ZstdDecoder::new(lfc_state.iter().as_slice())
+        .read_to_end(&mut uncompressed)
+        .await
+        .context("decoding LFC state")?;
+    let uncompressed_len = uncompressed.len();
+    info!(%uri, "downloaded LFC state, uncompressed size {uncompressed_len}, loading into postgres");
+
+    ComputeNode::get_maintenance_client(conn)
+        .await
+        .context("connecting to postgres")?
+        .query_one("select prewarm_local_cache($1)", &[&uncompressed])
+        .await
+        .context("loading LFC state into postgres")
+        .map(|_| ())
+}
+
+#[derive(Serialize, Default)]
+pub(in crate::http) struct PrewarmStatus {
+    status: compute_api::responses::PrewarmStatus,
+    error: String,
+    segments: i32,
+    done: i32,
+}
+
+// If prewarm failed, we want to get overall number of segments as well as done ones.
+// However, this route should be reliable even if querying postgres failed.
+pub(in crate::http) async fn prewarm_lfc_status(state: State) -> Json<PrewarmStatus> {
+    info!("requesting LFC prewarm status from postgres");
+    let mut status = PrewarmStatus::default();
+    {
+        let state = &state.state.lock().unwrap().prewarm_state;
+        status.status = state.status;
+        status.error = state.error.clone();
+    }
+
+    let res = match ComputeNode::get_maintenance_client(&state.tokio_conn_conf).await {
+        Ok(res) => res,
+        Err(err) => {
+            error!(%err, "connecting to postgres");
+            return Json(status);
+        }
+    };
+    let row = match res.query_one("select * from get_prewarm_info()", &[]).await {
+        Ok(row) => row,
+        Err(err) => {
+            error!(%err, "querying LFC prewarm status");
+            return Json(status);
+        }
+    };
+    status.done = row.try_get(0).unwrap_or_default();
+    status.segments = row.try_get(1).unwrap_or_default();
```

Comment:
nit: the default values might be misleading. I think it's better to make them optional in the status.

Reply:
This was requested in the RFC, so I'd better leave this as is.
https://github.com/neondatabase/neon/pull/11294